⬆️ UPGRADE: sphinx v3, myst-nb v0.10

chrisjsewell opened the pull request:
- [x] Add tox for sphinx 2/3 tests, and docs builds
- [x] add github CI for sphinx 2/3
- [x] add windows support (#575)
- [x] Add new options to `_config.yml` (including config for turning on all myst-parser extended syntaxes at once?)
- [x] Add more CLI options related to sphinx (nitpick, keep-going, freshenv)
- [x] verbosity CLI option? #585
- [x] document new CLI options
- [x] Document changes from myst-parser CHANGELOG
- [x] Document changes from myst-nb CHANGELOG
- [x] Check for documentation improvements to address issues in https://github.com/orgs/executablebooks/projects/5#card-44324532
  - [x] #613
  - [x] #875
  - [x] #878
  - [x] #762
  - [x] #745
  - [x] #828
  - [x] #754
  - [x] #743
  - [x] #694 
  - [x] #430 
  - [x] #705
  - [x] #629
  - [x] executablebooks/meta#114
  - [x] executablebooks/meta#111

- [x] Improve error reporting (don't show spurious traceback)
- [x] Improve testing (move from subprocess to click.CliRunner, fixes #896)
- [x] Lazy imports for click functions
- [x] Replace admonitions with new style
- [x] Replace nested bullet-lists with definition lists
- [x] add correct myst config to docs
- [x] include/document sphinx-panels #690
- [x] Add warning that Python 3.8 on Windows does not currently support notebook execution (#906)
- [x] add nbclient to versions reported in `jb --version`
- [x] Add note about sphinx extensions for images (#918)
- [x] add redirect for content-type -> file-types pages (using https://github.com/wpilibsuite/sphinxext-rediraffe)
- [x] Final look-through of docs
- [x] Add v0.8.0 changelog and release commit

- Post PR merge TODOs
  - [ ] close all referenced issues
  - [ ] "you could even have a role that standardises the format, e.g. {myst-role}`nb-exec-table` 
  - [ ] consider making xelatex the default
  - [ ] Make the conig schema the "source of truth" for available values and defaults, and auto-generate default_config from the schema
  - [ ] Update the jupytext version from ==1.6.0rc0 to ~=0.16.0, when it is released

chrisjsewell commented:
Hey @mmcky @AakashGfude you might want to start looking through this.
I've finished on the main code changes, so just need to work on the document updates.

You can look through the commit history to see what's changed.
This **will** be a merge commit lol, since all the commits are well defined/formatted, but also combine to achieve the overarching PR goal.

chrisjsewell commented:
> Hey mmcky AakashGfude you might want to start looking through this.

and also @choldgraf, If you have a spare moment in between the diaper changes 😉

mmcky commented on docs/contribute/intro.md:
````diff
@@ -44,19 +44,20 @@ code looks correct according to a few checks.
-(developer/tests)=
-## Testing infrastructure
+```shell
+>> pytest --ignore=tests/test_pdf.py
````
@chrisjsewell are we still ignoring `test_pdf`? I thought you fixed that recently on `CI` or is this the suggested way to run the tests locally to avoid dependencies on `tex`?

mmcky replied:
btw - I see the text below about LaTeX - but is that the recommended way to run tests locally you think?

chrisjsewell replied:
Yeh I think it works well enough, I guess the main thing is just to note that you need TeX installed to have certain tests pass.
But actually it would be better to decorate these tests with `@pytest.mark.requires_tex`, then you would run `pytest -k 'not requires_tex`

chrisjsewell replied:
I think I'll do this in a different PR, since its not a priority

mmcky commented on docs/advanced/sphinx.md:
````diff
@@ -169,11 +169,11 @@ You can also define TeX macros for a specific file by introducing them at the be
 ```md
 $$
-A = \bmat 1 & 1 \\ 2 & 1\\ 3 & 2 \emat,\ b=\bmat 2\\ 3 \\ 4\emat,\ \gamma = 0.5
+A = \bmat{} 1 & 1 \\ 2 & 1\\ 3 & 2 \emat{},\ b=\bmat{} 2\\ 3 \\ 4\emat{},\ \gamma = 0.5
````
@chrisjsewell this isn't usually how I would write this in latex. Typically macro substitutions I have used in the past don' thave `{}` -- is there a reason for adding these?

chrisjsewell replied:
Yes because what your doing is wrong lol. Look at the current documentation, and see how `\bmat` "eats up" the preceding 1: https://jupyterbook.org/advanced/sphinx.html
You need to close them to stop any arguments being passed

mmcky replied:
I see -- so the `array` environment is eating the `1` as an option -- gotcha. Macros I have used in the past have been more direct options (like operator substitutions). You're right!.

chrisjsewell replied:
FYI https://www.nongnu.org/chktex/ is a good linter for latex, which will complain about these "open" commands

mmcky commented on jupyter_book/commands/__init__.py:
```diff
@@ -50,23 +64,61 @@ def main():
     type=click.Choice(list(BUILDER_OPTS.keys())),
 )
-def build(path_source, path_output, config, toc, warningiserror, builder):
+def build(
```
given these changes to `build` I will hold off updating #687 until this is merged.

chrisjsewell replied:
Yeh thats why I wanted to give you the heads-up! so I don't introduce too many merge conflicts into your PRs lol

mmcky commented on jupyter_book/commands/__init__.py:
```diff
@@ -1,39 +1,53 @@
-"""Defines the commands that the CLI will use."""
+"""Defines the commands that the CLI will use.
+
+IMPORTANT: Top-level imports should be minimised here, to improve CLI responsiveness
```
are imports really that expensive? I have always thought that best practice / style was having the imports at the top of a file as a single point of reference.

mmcky replied:
if we do `lazy` imports is there any documentation we should add at the top of the module such as:

```
# {{ function }}:: {{ imports }}
```

chrisjsewell replied:
Normally, thats absolutely the case. But not for CLI, because all these imports have to install everytime you call the CLI and does have a noticeable difference. This is particularly accute, if you think about adding [tab completion](https://github.com/click-contrib/click-completion)

Lazy imports is also what we do in AiiDa: https://github.com/aiidateam/aiida-core/tree/develop/aiida/cmdline/commands

chrisjsewell replied:
> if we do lazy imports is there any documentation 

Personally, I don't think its necessary, and also it would be prone to becoming outdated.
The main thing with the imports, is that we check they all work, which is achieved by the good test coverage

mmcky replied:
Gotcha. I guess in that case I am just wondering how we determine first point of entry for an import? Or do you specify what is needed in each function?

mmcky replied:
Do `imports` respect function scope or are they still globally available?

chrisjsewell replied:
> first point of entry for an import?

not quite sure what you mean by this? If you import the same thing multiple times, that's not a problem since it caches them

you just have to make sure that things are available within the scope you are working, which flake8 will warn about any if something is not available (+ would be picked up in the tests)

mmcky reviewed (commented):
Thanks @chrisjsewell these are nice additions to `jupyter-book`. That validation of `_config.yml` will be handy.

chrisjsewell commented:
> That validation of _config.yml will be handy.

Cheers 😄 I have just made it a warning for now, rather than failing, because I didn't want to have a mistake in the schema lead to everyones builds starting to error

mmcky commented:
> I have just made it a warning for now

I think that is the best way to go 👍

chrisjsewell commented:
@mmcky @AakashGfude in 6cff10d I have just fixed how the `_config.yml` is located for single page builds, and added some information printing 😄 

So for a project build:

![image](https://user-images.githubusercontent.com/2997570/91375186-e4447080-e811-11ea-9145-b10013899196.png)

and for a page build:

![image](https://user-images.githubusercontent.com/2997570/91375170-d8f14500-e811-11ea-9475-75febc16140a.png)

Let me know if you see any issues with this

mmcky commented:
@chrisjsewell nice -- the path information will be really useful.

AakashGfude commented on docs/contribute/intro.md:
````diff
@@ -44,19 +44,20 @@ code looks correct according to a few checks.
 ```
 
-This will run the Jupyter Book test suite, *except for the PDF tests*. This is because
+Either will both run the Jupyter Book test suite, *except for the PDF tests*. This is because
````
Just being pedantic here :sweat_smile: . Should we start `This is because` in a new line?

chrisjsewell replied:
Yeh not going through and "fixing" all the one-line pre sentence issues 😆 but I have done a bit of tidying in 8511e86, including this

AakashGfude commented on jupyter_book/default_config.yml:
```diff
@@ -34,6 +46,11 @@ html:
+#######################################################################################
+# LaTeX-specific settings
+latex:
+  latex_engine              : pdflatex
```
are we sticking with pdflatex or xelatex? xelatex is a bit slower, but has support for wider range of characters which can come handy.

chrisjsewell replied:
Yeh I would say xelatex. Is it always packaged with TeX distributions though?
We can also add a comment string here listing the alternatives: https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-latex_engine

AakashGfude replied:
cc: @mmcky

mmcky replied:
`pdflatex` is the default used by `sphinx` so I imagine most testing in `sphinx` is against `pdflatex` (as the default option). We always use `xelatex` on `quantecon` projects -- and is a good default choice for more scientific book. But there is also `lualatex` etc. 

It isn't hard to set in `_config.yml` particularly if we setup the default value so that `jb create` populates with an easy to change setting. 

Essentially I don't really mind either way -- I think `pdflatex` isn't a bad "industry standard" but I can't think of a case where building with `xelatex` will be any worse off than `pdflatex`.

mmcky replied:
re: testing -- come to think of it most tests are probably just against `tex` output so there may not be a lot of testing against the pdf conversion method.

chrisjsewell replied:
We could just set it to None, to use the sphinx default, then yeh add a commment with the main options

AakashGfude replied:
@chrisjsewell yes it is available with all the major TeX distributions like TeX Live, MikTex, MacTex, Lyx. Not sure if any other distributions are common though.  https://en.wikipedia.org/wiki/XeTeX .
Hmm. True, we can list the alternatives, but hopefully, the theme we create for LaTex does not behave weirdly in other alternatives. Only time can tell.

mmcky replied:
theme support is a good point @AakashGfude. Lets visit that in #586 

@chrisjsewell I don't see a downside in keeping it `pdflatex` rather than `None` as it is more descriptive and explicit in terms of what is being used.

chrisjsewell replied:
Fair, I don't see it as a big issue, although I do note that sphinx use xelatex as the default for chinese and greek.

mmcky replied:
I guess in summary I think `pdflatex` is a pretty sane default value  -- despite the fact that I will need to update it to `xelatex` for all our projects `:-)`. I think "most" users of `tex` interface with `pdflatex` by default -- but a good writeup of options will be really useful.

mmcky replied:
OK - well then let's be brave and set `xelatex` for built-in unicode support. It is getting more and more popular.

chrisjsewell replied:
I might do this in a separate PR, since I have a feeling it may break the CI
(I've added a comment to the default_config with the options)

RojerGS replied:
Just want to say that there are people out there (e.g. me) for whom `pdflatex` doesn't cut it because of its poor support for Unicode.

chrisjsewell commented:
Ok @mmcky @AakashGfude @choldgraf @najuzilu pick the bones out of this!
I've basically "refreshed" every page in the documentation 😆

mmcky commented on docs/advanced/sphinx.md:
```diff
@@ -138,11 +140,22 @@ you configure in `conf.py`. To do so, use the following section of `_config.yml`
 
+### Fine control of parsing and execution
+
+As discussed in [the components of Jupyter Book](intro/jupyter-book-components), two of the principe components of Jupyter Book are sphinx extensions;
```
```suggestion
As discussed in [the components of Jupyter Book](intro/jupyter-book-components), two of the principle components of Jupyter Book are sphinx extensions;
```

mmcky commented on docs/content/content-blocks.md:
```diff
@@ -110,6 +77,69 @@ My content
+The admonition syntax above utilises the general [directives syntax](content:myst/directives).
+This is has the advantage of making it consistent with every other directive.
+However, a big disadvantage is that, when working in any standard Markdown editor (or the Jupyter Notebook interfaces),
+the text they contain will not render nicely as standard Markdown.
```
@chrisjsewell just to clarify -- this is for compatibility with local previews for `editing` right (or jupyter notebook)? The usual "```" directive should now contain `md` -- it is just that editors won't syntax highlight them as they are interpreted as literal (with a highlight definition).

chrisjsewell replied:
> this is for compatibility with local previews for editing right (or jupyter notebook)?

Yep thats correct

> The usual "```" directive should now contain `md`

Not sure I get exactly what you mean by this?
But essentially, under the hood, it is conceptually different; because instead of passing the "raw" text to a sphinx directive (and getting back a docutils node), you are taking Markdown tokens and constructing the docutils nodes from that

chrisjsewell replied:
and yeh, as explained, I think its just a lot more new user friendly.
For example in this page the first admonition now correctly shows the URL links https://github.com/executablebooks/jupyter-book/blob/upgrade/myst-nb-0.9/docs/intro.md

mmcky replied:
thanks @chrisjsewell I think I just added a suggestion to emphasis local previews to prevent possible confusion with parsing via `jupyter-book`

mmcky commented on docs/content/content-blocks.md:
```diff
@@ -110,6 +77,69 @@ My content
+The admonition syntax above utilises the general [directives syntax](content:myst/directives).
+This is has the advantage of making it consistent with every other directive.
+However, a big disadvantage is that, when working in any standard Markdown editor (or the Jupyter Notebook interfaces),
+the text they contain will not render nicely as standard Markdown.
```
```suggestion
the text they contain will not render nicely as standard Markdown (for local Markdown previews).
```

chrisjsewell replied:
I'm not sure if "local" is quite the right word here. As I mention above, it also makes the pages a lot nicer to view on GitHub, or basically anywhere where you are directly viewing the source Markdown file

mmcky replied:
Roger that -- perhaps "(for Markdown previews)" to emphasis this is for preview mechanisms rather than build output.

mmcky commented on docs/content/content-blocks.md:
```diff
@@ -110,6 +77,69 @@ My content
+:::
+::::
+
+This syntax only supports a select subset of directives:
+
+> admonition, attention, caution, danger, error, important, hint, note, seealso, tip and warning.
```
```suggestion
:::{note}
This syntax only supports a select subset of directives:

> admonition, attention, caution, danger, error, important, hint, note, seealso, tip and warning.
:::
```

mmcky replied:
I think a note admonition would stand out a bit more in our theme.

mmcky commented on docs/content/content-blocks.md:
````diff
@@ -150,16 +180,99 @@ results in:
+  myst_extended_syntax: true
+```
+
+Definition lists utilise the [markdown-it-py deflist plugin](https://markdown-it-py.readthedocs.io/en/latest/plugins.html), which itself is based on the [Pandoc definition list specification](http://johnmacfarlane.net/pandoc/README.html#definition-lists).
````
@chrisjsewell do we support any special styling in the theme `sphinx-book-theme`? I guess I am curious how we do modular additions to `themes` such as if we did include styling.

chrisjsewell replied:
See https://github.com/executablebooks/sphinx-book-theme/pull/191, although I'm open to improvements, since I literally added it a few days precisely to get this to work 😄

chrisjsewell replied:
basically this is the current SCSS:

```scss
        dl.simple {
            & dd {
                margin-left: 1.5em;
            }
            & dd:not(:last-child) {
                margin-bottom: 0px;
                p:last-child {
                    margin-bottom: 0px;
                }
            }
        }
```

mmcky replied:
Cool. So essentially a theme needs to support all the elements -- and if they are present they get styled -- otherwise it doesn't get used. So this `SCSS` doesn't get added to the theme (via plugin style code injection) if definition lists are used.I don't see this as a problem -- just understanding how themes work with `modularity`.

chrisjsewell replied:
Yeh I guess basically the key to good theme/CSS modularity is having lots of classes on the HTML elements, so its easier to precisely specify things.

mmcky commented on docs/content/content-blocks.md:
```diff
@@ -173,13 +286,12 @@ Here's my figure:
 
-See {doc}`glue` for more information on how to use Glue to insert your outputs
-directly into your content.
+See [](content:code-outputs:glue) for more information on how to use Glue to insert your outputs directly into your content.
```
The `[]` takes the title from the target document right?

chrisjsewell replied:
yep 👍

chrisjsewell replied:
Should maybe document that in citations.md

mmcky replied:
That sounds good -- I didn't know you could do that.

chrisjsewell replied:
added 👍

mmcky commented on docs/content/content-blocks.md:
`````diff
@@ -233,44 +345,21 @@ Here is a cool quotation.
 
-````
+````md
 ```{glossary}
`````
@chrisjsewell a `glossary` is using syntax required by the directive right? We can't use definition lists here.

chrisjsewell replied:
correct; as with all ```` ``` ```` directives, the content is not initially parsed to Tokens by the markdown parser, but simply supplied directly to the directive (after extracting arguments/options)

mmcky commented on docs/content/figures.md:
```diff
@@ -1,11 +1,74 @@
+
+<img src="../images/fun-fish.png" alt="fishy" class="bg-primary mb-1" width="200px">
+
+This will also be output in PDF LaTeX builds!
```
@chrisjsewell no way! -- nicely done. That's cool. Is this in the parser to unpack `html`?

chrisjsewell replied:
yep also, as of today, see https://myst-parser.readthedocs.io/en/latest/using/syntax-optional.html#markdown-figures, which allows you to also add a reference target and a caption 😃

chrisjsewell replied:
This is the HTML parser code: https://github.com/executablebooks/MyST-Parser/blob/master/myst_parser/parse_html.py
actually pretty simple

chrisjsewell replied:
the `:::` blocks are basically a new "breed" of Markdown specific directives.
As I plan to add more (see https://github.com/executablebooks/MyST-Parser/issues/213), I will try to formalise their usage, so that they are more modular and extensible

mmcky commented on docs/content/myst.md:
````diff
@@ -1,55 +1,52 @@
 ```{margin}
 For those who are familiar with Sphinx, MyST markdown is basically
-CommonMark + Sphinx roles and directives
+CommonMark + MArkdown Extensions + Sphinx roles and directives
````
```suggestion
CommonMark + Markdown Extensions + Sphinx roles and directives
```

mmcky commented on docs/contribute/intro.md:
```diff
@@ -139,8 +138,8 @@ what kinds of functionality they support:
-  controls some parts of notebook execution. It also provdes
-  {doc}`inserting code outputs into content <../content/glue>`.
+  controls some parts of notebook execution.
+  It also provdes [inserting code outputs into content](content:code-outputs:glue).
```
```suggestion
  It also provides [inserting code outputs into content](content:code-outputs:glue).
```

mmcky commented on docs/file-types/jupytext.md:
```diff
@@ -0,0 +1,99 @@
+sphinx:
+  config:
+    nb_custom_formats:
+        .mysuffix: mylibrary.converter_functio
```
```suggestion
        .mysuffix: mylibrary.converter_function
```

mmcky commented on docs/start/overview.md:
```diff
@@ -188,7 +189,7 @@ You can store these files in whatever collection of folders you'd like, note tha
 items in your `_toc.yml` file (see below section)
 
-
+()
```
Should this contain reference text?

chrisjsewell replied:
good spot

mmcky reviewed (commented):
Love your work @chrisjsewell -- just some minor comments. No doubt you've responded to my review points already so I will cycle back and iterate :-)

chrisjsewell commented on docs/start/overview.md:
```diff
@@ -188,7 +189,7 @@ You can store these files in whatever collection of folders you'd like, note tha
 items in your `_toc.yml` file (see below section)
 
-
+()
```
```suggestion

```

chrisjsewell commented:
> Love your work @chrisjsewell -- just some minor comments. No doubt you've responded to my review points already so I will cycle back and iterate :-)

Thanks @mmcky! now incorporated

chrisjsewell commented:
Markdown Figures! https://deploy-preview-911--jupyter-book.netlify.app/content/figures.html#markdown-figures

phaustin commented:
@chrisjsewell -- looking good!  appreciate the pytest master class

chrisjsewell commented:
well you guys laid the ground work 👍

choldgraf commented on docs/content/code-outputs.md:
```diff
@@ -204,7 +366,7 @@ control over how the outputs look in your pages.
 +++
 
-## Controling the pasted outputs
+### Controling the pasted outputs
```
```suggestion
### Controlling the pasted outputs
```

choldgraf commented on docs/content/content-blocks.md:
`````diff
@@ -66,6 +23,16 @@ Results in the following output:
 Here is a note!
 ```
 
+````{margin} A note on nesting
`````
perhaps margins/sidenotes should be `::::` as well?

chrisjsewell replied:
Way ahead of you 😄 https://github.com/executablebooks/MyST-Parser/issues/213
As I mentioned to @mmcky (https://github.com/executablebooks/jupyter-book/pull/911#discussion_r478763053):

the `:::` blocks are basically a new "breed" of Markdown specific directives (i.e. they require as input Markdown-It Tokens, rather than raw text).

As I plan to add more (see https://github.com/executablebooks/MyST-Parser/issues/213), I will try to formalise their usage, so that they are more modular and extensible
(currently they are a bit hard-coded into the renderer https://github.com/executablebooks/MyST-Parser/blob/bf7fd80c3273a4fea4be595bfb37a5b51b452760/myst_parser/docutils_renderer.py#L676)

choldgraf commented on docs/content/content-blocks.md:
```diff
@@ -110,6 +77,72 @@ My content
+### New Style Admonitions
+
+The admonition syntax above utilises the general [directives syntax](content:myst/directives).
+This is has the advantage of making it consistent with every other directive.
```
```suggestion
This has the advantage of making it consistent with every other directive.
```

choldgraf commented on docs/content/content-blocks.md:
```diff
@@ -150,16 +183,99 @@ results in:
+Term 2
+: Definition
+
+From the Pandoc documentation:
```
```suggestion
From the [Pandoc documentation](https://pandoc.org/MANUAL.html#definition-lists):
```

choldgraf commented on docs/content/execute.md:
```diff
@@ -119,32 +115,88 @@ execute:
+:::
 
-### How can I include code that raises errors?
+By default, the command working directory (cwd) that a notebook runs in will be the directory it is located in (for both `auto` and `cache`). This means that notebooks requiring access to assets in relative paths will work.
```
```suggestion
By default, the command working directory (cwd) in which a notebook runs will be the directory in which it is located (for both `auto` and `cache`). This means that notebooks requiring access to assets in relative paths will work.
```

choldgraf commented on docs/content/execute.md:
```diff
@@ -161,3 +213,86 @@ tags: [raises-exception]
+## Execution statistics
+
+As notebooks are executed, certain statistics are stored on the build environment by MyST-NB.
+The simplest way to access and visualise this data is using the `nb-exec-table` directive.
```
```suggestion
The simplest way to access and visualise this data is using the **`{nb-exec-table}`** directive.
```

I saw you did this elsewhere and kinda liked it...maybe we want to adopt that as a standard way to refer to directives in these docs?

chrisjsewell replied:
sounds reasonable; you could even have a role that standardises the format, e.g. ``{myst-role}`nb-exec-table` ``

choldgraf commented on docs/content/figures.md:
```diff
@@ -1,15 +1,78 @@
+This will correctly copy the image to the build folder and will render it in all output formats (HTML, TeX, etc).
+However, it is limited in the configuration that can be applied, for example setting a width.
+
+As discussed in [this section](content:myst/directives), MyST allow for directives to be used such as `image` and `figure` (see [the sphinx documentation](sphinx:rst-primer) for available options):
```
```suggestion
As discussed in [this section](content:myst/directives), MyST allows for directives to be used such as `image` and `figure` (see [the sphinx documentation](sphinx:rst-primer) for available options):
```

choldgraf commented on docs/content/figures.md:
```diff
@@ -1,15 +1,78 @@
+This allows you to control aspects of the image with [directive arguments](directive-arguments).
+
+In one way, this is an improvement on the Markdown syntax, however, the drawback is that this syntax will not show the image in common Markdown viewers (for example when the files are viewed on GitHub).
+The final option then is directly using HTML, which can also parsed by MyST Markdown.
```
```suggestion
The final option then is directly using HTML, which can also be parsed by MyST Markdown.
```

choldgraf commented on docs/content/figures.md:
```diff
@@ -1,15 +1,78 @@
+
+In one way, this is an improvement on the Markdown syntax, however, the drawback is that this syntax will not show the image in common Markdown viewers (for example when the files are viewed on GitHub).
+The final option then is directly using HTML, which can also parsed by MyST Markdown.
+U
```
accidental line break?

choldgraf commented on docs/content/figures.md:
```diff
@@ -18,27 +81,86 @@ Here is my figure caption!
+- {ref}`directive-fig`
+- [](markdown-fig)
+
 Another convenient way to create cross-references is with the `{numref}` role,
```
I think we should have a header that has "numbered references" or something, so people find it more easily

chrisjsewell replied:
added in coming commit

choldgraf commented on docs/file-types/jupytext.md:
```diff
@@ -0,0 +1,99 @@
+sphinx:
+  config:
+    nb_custom_formats:
+        .ipynb:
```
this is really cool

choldgraf commented on docs/file-types/jupytext.md:
```diff
@@ -0,0 +1,99 @@
+        .ipynb:
+            - nbformat.reads
+            - as_version: 4
+            - true
```
I wonder if we should make this a key:val argument just to make it easier for users to know what they're configuring. e.g. `commonmark: true` rather than just `true`. WDYT?

chrisjsewell replied:
Yeh its currently adapted directly from https://nbsphinx.readthedocs.io/en/0.7.1/custom-formats.html.
Does it sound reasonable to add `custom_formats` to the `parse` key, and do the transform in `yaml_to_sphinx`, rather than changing it in `myst-nb`?

chrisjsewell replied:
??

choldgraf commented on docs/file-types/notebooks.ipynb:
```diff
@@ -6,20 +6,18 @@
-    "[CommonMark Markdown](https://commonmark.org/) with minor modifications.\n",
-    "For more information about writing Jupyter-flavored markdown in Jupyter Book,\n",
-    "see {doc}`markdown`.\n",
+    "This is mostly a flavor of markdown called [CommonMark Markdown](https://commonmark.org/) with minor modifications.\n",
```
```suggestion
    "This is mostly a flavour of markdown called [CommonMark Markdown](https://commonmark.org/) with minor modifications.\n",
```

;-)

choldgraf commented on jupyter_book/config_schema.json:
```diff
@@ -0,0 +1,212 @@
+{
+    "$schema": "http://json-schema.org/schema#",
```
this is also amazing

choldgraf commented on jupyter_book/default_config.yml:
```diff
@@ -15,8 +15,19 @@ exclude_patterns            : []  # Patterns to skip when building the book. Can
   execute_notebooks         : auto  # Whether to execute notebooks at build time. Must be one of ("auto", "force", "cache", "off")
-  cache                     : ""  # A path to the jupyter cache that will be used to store execution artifacs. Defaults to `_build/.jupyter_cache/`
-  exclude_patterns          : []  # A list of patterns to *skip* in execution (e.g. a notebook that takes a really long time)
+  cache                     : ""    # A path to the jupyter cache that will be used to store execution artifacs. Defaults to `_build/.jupyter_cache/`
```
can we auto-generate this page from the schema at some point? (not this PR)

chrisjsewell replied:
Yep sounds like a good idea, then the shchema would be the "source of truth" for available values and defaults

choldgraf replied:
yeah exactly

chrisjsewell replied:
Added to post PR merge TODOs (see initial PR comment)

choldgraf commented on jupyter_book/sphinx.py:
```diff
@@ -84,7 +89,7 @@ def build_sphinx(
 
     # Start with the default Sphinx config
-    sphinx_config = DEFAULT_CONFIG.copy()
+    sphinx_config = get_default_config()
```
just curious - why is this a better practice than copying the dict?

chrisjsewell replied:
Mutable global variable are BAD! Its one of my grievances with docutils/sphinx, because it makes them not thread-safe and thus incompatible with asyncio

Basically any mutuations to DEFAULT_CONFIG leads to unreliable results during multiple calls, which is what we found is occuring in #896

choldgraf commented on jupyter_book/yaml.py:
```diff
@@ -63,6 +99,19 @@ def yaml_to_sphinx(yaml):
+    # parse and rendering
+    parse = yaml.get("parse")
+    if parse:
+        if parse.get("myst_extended_syntax") is True:
```
I wonder if we should have a single section that lists all of the things that are activated if `myst_extended_syntax` is true. We refer to it throughout the docs but in a piecemeal fashion.

chrisjsewell replied:
Well that single section is essentially: https://myst-parser.readthedocs.io/en/latest/using/syntax-optional.html. 
This is referred to in the "Other MyST Markdown syntax" heading in docs/content/myst.md.
Should I make it more explicit?

choldgraf replied:
let's just leave it for now and see if it creates confusion in folks over time, I think it's good 👍

choldgraf commented:
Hey all - some comments above! One thing that we should discuss at some point is what we want to treat as "canonical MyST". As we add new syntactic sugar and features (like `:::`) and as more applications/libraries support MyST, it'll be more important to have a _standard_ that we guarantee will be treated the same across all these applications. Right now I don't think it's such a big deal, but as we have more and more flags for "turn on support for XYZ syntax" I think it'll become more important.

chrisjsewell commented:
> I have a few nits in there, some comments/questions/etc, but in general this is great and I think close to merging.

Ah brilliant thanks @choldgraf, I was hoping you could find the time to review, before I completely rewrote your "baby" (pin intended 👶)

I'm just half-way through changing `yaml.py` (and adding regression tests) to address this issues #915, the basic premise being that `default_config.yml` should always actually be the default, unless you specifically override something.

Then the plan is   

- go through and address your comments
- add sphinx panels to dependencies and document it
- Add the release commit directly to this PR (version -> 0.8.0 and write changelog.md), so then I can release it immediately after merging, so as to reduce the time between the documentation being out of sync with the package. This is something that I noted with @AakashGfude is a big drawback of currently not being able to use ReadTheDocs; which can store multiple versions. Perhaps somehow we should add the version the documentation was built from just below the logo, so that people are not confused by things not working as documented if they haven't yet upgraded jupyter-book?

chrisjsewell commented on docs/intro.md:
```diff
@@ -80,7 +79,7 @@ These files are written in [MyST markdown](content/myst), an
 They could alternatively have been written directly as Jupyter notebooks.
 
-**Build the demo book**
+### Build the demo book
```
Ah yes, I changed places where this was the case, then realised there was probably a reason for it and changed them back (but missed this one).
Is this a bug then; something we should fix or warn about in the documentation?

chrisjsewell replied:
```suggestion
**Build the demo book**
```

choldgraf replied:
not quite a bug - more like "annoying but expected behavior according to Sphinx". We do mention it here: https://jupyterbook.org/customize/toc.html#how-headers-and-sections-map-onto-to-book-structure

chrisjsewell commented on docs/_config.yml:
```diff
@@ -10,6 +10,9 @@ description: >- # this means to ignore newlines until "baseurl:"
 execute:
   execute_notebooks: cache
 
+parse:
```
Yeh I just thought a `parse` key was more general/user-friendly, to directly infer what its purpose is

choldgraf replied:
parse works too 👍 just wanted to throw out myst in case others thought it was a good idea

chrisjsewell commented on docs/_toc.yml:
```diff
@@ -10,13 +10,13 @@
   - file: customize/config
   - file: customize/toc
-  - file: content-types/index
+  - file: file-types/index
```
So shall we try adding executablebooks/meta#117 then?
As a "new developer" to JB I just found content-types to be very confusing, since you also have the `content` folder which more accurately is a section on content types

choldgraf replied:
yeah I agree the renaming makes more sense 👍 

maybe we could try out the extension and see if it is straightforward to get working?

chrisjsewell replied:
Works, but two annoyances: https://github.com/wpilibsuite/sphinxext-rediraffe/pull/17, https://github.com/wpilibsuite/sphinxext-rediraffe/issues/18

chrisjsewell replied:
These will be merged tomorrow and a new version released 😄

chrisjsewell commented on docs/advanced/advanced.md:
```diff
@@ -71,9 +71,9 @@ This will entirely remove the folders in the `_build/` directory.
+## How should I add cell tags and metadata to my notebooks?
 
-You can control the behavior of Jupyter Book by putting custom tags
+You can control the behaviour of Jupyter Book by putting custom tags
```
Yep this is my house now 😆

chrisjsewell commented on docs/advanced/pdf.md:
````diff
@@ -92,18 +92,21 @@ For `Windows` please install [texlive](https://www.tug.org/texlive/windows.html)
 ```
 
-```{note}
+::::{note}
````
No we don't, they are parsed as two distinct entities in markdown-it-py. Although I don't think it hurts

chrisjsewell commented:
> One thing that we should discuss at some point is what we want to treat as "canonical MyST". As we add new syntactic sugar and features (like :::) and as more applications/libraries support MyST, it'll be more important to have a standard that we guarantee will be treated the same across all these applications. Right now I don't think it's such a big deal, but as we have more and more flags for "turn on support for XYZ syntax" I think it'll become more important.

Yeh thats maybe the tricky part of having a "configurable syntax", I guess Pandoc has the same issue.

One thing I did do in myst-parser is to split off:

- https://myst-parser.readthedocs.io/en/latest/using/syntax.html
- https://myst-parser.readthedocs.io/en/latest/using/syntax-optional.html

To make it more clear what is and is not enabled by default.

Certainly I feel that the `:::` admonitions are massively more user-friendly when editing/viewing source files, particularly in a Jupyter Notebook context. So I would like to push those as the *de facto* standard

chrisjsewell commented on docs/content/figures.md:
```diff
@@ -1,15 +1,78 @@
+
+In one way, this is an improvement on the Markdown syntax, however, the drawback is that this syntax will not show the image in common Markdown viewers (for example when the files are viewed on GitHub).
+The final option then is directly using HTML, which can also parsed by MyST Markdown.
+U
+sing raw HTML is usually a bad choice (see [this explanation](raw-html-in-markdown)),
```
```suggestion
Using raw HTML is usually a bad choice (see [this explanation](raw-html-in-markdown)),
```

chrisjsewell commented:
Another day another big change lol.
See the bb0b8d7 commit message for details.

@mmcky @choldgraf wanna take a look?
Also @RojerGS do you want to look and see if this fixes #915?
Particularly, are the generated files in tests/test_config as expected and/or should any more tests be added?

RojerGS commented:
> Also @RojerGS do you want to look and see if this fixes #915?

I really want to help but I'll be AFK for a week and I don't want to slow you down. The changes _look_ good but I didn't test them.
 
> I think the expected behaviour still needs to be clarified/discussed a bit more though

In what sense? Do we agree that the defaults that will be in the docs should be the values used unless we explicitly set that specific `key: value` pair, regardless of its depth in the config dictionary?
Also, does 5644c27 fully reset a dict to the sphinx defaults or to the jb defaults? Either way, I think I missed the point of allowing such a thing to happen, it feels like it will make the config behaviour more complicated and with more special cases.

chrisjsewell commented:
> I really want to help but I'll be away from a computer for a week and I don't want to slow you down. The changes look good but I didn't test them.

Yeh no problem, just wanted to give you the heads up 😄 

> In what sense? Do we agree that the defaults that will be in the docs should be the values used unless we explicitly set that specific key: value pair, regardless of its depth in the config dictionary?

Well mainly I just want someone(s) else to give it a second look, to make sure I'm not going to inadvertently break everyone's existing books 😬 
Personally, I think the new logic is reasonable and, at the very least, we now have a way to "programmatically document" the behaviour (*via* the test cases), and also a way for users to directly understand how their config is being translated (*via* `jb config sphinx`).

> it feels like it will make the config behaviour more complicated and with more special cases

Agree, I have removed this in c01f1b2.

Now the only "special case" is that anything set in `sphinx: config: ...` will directly override anything set previously (in sphinx/yaml defaults), rather than recursively updating.
(see e.g. tests/test_config/test_get_final_config_sphinx_.yml).

chrisjsewell commented:
Ok, this is nearly done; just the addition of https://github.com/wpilibsuite/sphinxext-rediraffe, and writing the v0.8.0 CHANGELOG

I'm looking to merge this and release v0.8.0 in the next ~24 hours.
So speak now or forever hold your peace!

phaustin commented:
reminder that we can remove  https://github.com/executablebooks/jupyter-book/blob/upgrade/myst-nb-0.9/docs/advanced/advanced.md#working-on-windows

chrisjsewell commented:
> reminder that we can remove [working-on-windows](https://github.com/executablebooks/jupyter-book/blob/upgrade/myst-nb-0.9/docs/advanced/advanced.md?rgh-link-date=2020-08-30T12%3A23%3A20Z#working-on-windows)

Actually, in 0cf6f69a29ec1f2b81ea6806a10813a494a388a1, I have already modified this; to say that we do now support/test against Windows OS 😄 . But noted the notebook execution issue with Python 3.8 still remains

chrisjsewell commented:
FYI also nearly fixed the jupytext issue, so no more myst-parser version incompatabilities or need for `jupytext[myst]` 😄 : https://github.com/mwouts/jupytext/pull/602

chrisjsewell commented:
and looks like issues with the logger 😬

choldgraf commented:
I'm +1 on giving this a merge if others are 👍 as well. This is a big one and probably better to get it in sooner than later so people have a chance to play with it and see if it breaks things. I guess we should also cut a release relatively quickly since the docs will be even more outdated w/ the latest release once it's merged?

chrisjsewell commented:
>  I guess we should also cut a release relatively quickly since the docs will be even more outdated w/ the latest release once it's merged?

No that what I meant in https://github.com/executablebooks/jupyter-book/pull/911#issuecomment-683178248; I will literally add the "release" commit to this PR, merge, then immediately release v0.8

chrisjsewell commented:
> have a chance to play with it and see if it breaks things

If you want to do that, you need to do it **now** on this PR

choldgraf commented:
Makes sense. Personally there's nothing I'd do other than to look at how the docs work,and they seem to be good to me. Maybe @mmcky can try this branch w the quantecon books? Perhaps @bmcfee can try with his book as well to make sure there's nothing unexpected

chrisjsewell commented:
Yep sounds good 👍

chrisjsewell commented:
The jupytext page is now actually written as an RMarkdown file 😄 https://deploy-preview-911--jupyter-book.netlify.app/file-types/jupytext.html#file-types-custom

chrisjsewell commented:
@AakashGfude @mmcky you may wish to check out https://deploy-preview-911--jupyter-book.netlify.app/content/figures.html#supported-image-formats, which is of relevance for LaTeX builds

mmcky commented:
> @AakashGfude @mmcky you may wish to check out https://deploy-preview-911--jupyter-book.netlify.app/content/figures.html#supported-image-formats, which is of relevance for LaTeX builds

thanks @chrisjsewell -- the `html` parsing is a nice feature as it brings it closer to compatibility with how `jupyter` notebooks get rendered in the notebook setting.

chrisjsewell commented:
Now thats what I call a chage log: https://deploy-preview-911--jupyter-book.netlify.app/reference/_changelog.html

RojerGS commented:
> Well mainly I just want someone(s) else to give it a second look, to make sure I'm not going to inadvertently break everyone's existing books 😬
> Personally, I think the new logic is reasonable and, at the very least, we now have a way to "programmatically document" the behaviour (_via_ the test cases), and also a way for users to directly understand how their config is being translated (_via_ `jb config sphinx`).

A bit late, but looks good!
